frontend/src/metabase/visualizations/echarts/pie/model/index.ts

```diff
@@ -0,0 +1,206 @@
+import { pieNegativesWarning } from "metabase/visualizations/lib/warnings";
+import { getDefaultColors } from "metabase/visualizations/shared/settings/pie";
+import type {
+  ComputedVisualizationSettings,
+  RenderingContext,
+} from "metabase/visualizations/types";
+import type { RawSeries, RowValue } from "metabase-types/api";
+
+import type { ShowWarning } from "../../types";
+import { OTHER_SLICE_MIN_PERCENTAGE, OTHER_SLICE_KEY } from "../constants";
+
+import type {
+  PieColumnDescriptors,
+  PieChartModel,
+  PieSliceData,
+} from "./types";
+
+function getColDescs(
+  rawSeries: RawSeries,
+  settings: ComputedVisualizationSettings,
+): PieColumnDescriptors {
+  const [
+    {
+      data: { cols },
+    },
+  ] = rawSeries;
+
+  const dimension = findWithIndex(
+    cols,
+    c => c.name === settings["pie.dimension"],
+  );
+  const metric = findWithIndex(cols, c => c.name === settings["pie.metric"]);
+
+  if (!dimension.item || !metric.item) {
+    throw new Error(
+      `Could not find columns based on "pie.dimension" (${settings["pie.dimension"]}) and "pie.metric" (${settings["pie.metric"]}) settings.`,
+    );
+  }
+
+  return {
+    dimensionDesc: {
+      index: dimension.index,
+      column: dimension.item,
+    },
+    metricDesc: {
+      index: metric.index,
+      column: metric.item,
+    },
+  };
+}
+
+export function getRowValues(row: RowValue[], colDescs: PieColumnDescriptors) {
+  const { dimensionDesc, metricDesc } = colDescs;
+
+  const dimensionValue = row[dimensionDesc.index];
+
+  const metricValue = row[metricDesc.index] ?? 0;
+  if (typeof metricValue !== "number") {
+    throw new Error(
+      `Pie chart metric value (${metricValue}) should be a number`,
+    );
+  }
+
+  return { dimensionValue, metricValue };
+}
+
+export function getPieChartModel(
+  rawSeries: RawSeries,
+  settings: ComputedVisualizationSettings,
+  renderingContext: RenderingContext,
+  showWarning?: ShowWarning,
+): PieChartModel {
+  const [
+    {
+      data: { rows },
+    },
+  ] = rawSeries;
+  const colDescs = getColDescs(rawSeries, settings);
+
+  // We allow negative values if every single metric value is negative or 0
+  // (`isNonPositive` = true). If the values are mixed between positives and
+  // negatives, we'll simply ignore the negatives in all calculations.
+  const isNonPositive = rows.every(
+    row => getRowValues(row, colDescs).metricValue <= 0,
+  );
+
+  const total = rows.reduce((currTotal, row) => {
+    const metricValue = getRowValues(row, colDescs).metricValue;
+    if (!isNonPositive && metricValue < 0) {
+      showWarning?.(pieNegativesWarning().text);
+      return currTotal;
+    }
+
+    return currTotal + metricValue;
+  }, 0);
+
+  // sometimes viz settings are malformed and "pie.colors" does not contain a
+  // key for the current dimension value, so we need to compute defaults to
+  // ensure every key has a color
+  const defaultColors = getDefaultColors(rawSeries, settings);
```

Comment:
I believe it is still better to keep viz settings related logic in settings definitions. I see there is `"pie._colors"` setting which is computed on fly based on the `settings["pie.colors"]`. Would using it work? If not, we could also define getValue for "pie._colors" and compute colors there so we don't have to do it here

Reply:
I think having two of the same setting (`pie.colors` and `pie._colors`) is an anti-pattern. It's not clear to developers which one should be used, and bugs will be created by using the wrong setting and not realizing it, or from the two settings diverging in values. We should avoid implementation details that depend on implicit knowledge instead of explicitly clear values in code.

What we should instead do is implement `getValue` for `pie.colors` alone.

Reply:
Ended up moving this logic to `getValue` for `pie.colors`, using the same underlying function both in `chart-definition` and the static viz default computation.